### PR TITLE
wal: check out of range slice in "ReadAll", entry limits in "decodeRecord"

### PR DESCRIPTION
**Pull request by gyuho**
Signed-off-by: Gyuho Lee <leegyuho@amazon.com>

Please read https://github.com/etcd-io/etcd/blob/master/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: wal/decoder.go [hunk 1]
```diff
@@ -79,6 +80,9 @@ func (d *decoder) decodeRecord(rec *walpb.Record) error {
 	}
 
 	recBytes, padBytes := decodeFrameSize(l)
+	if recBytes >= math.MaxInt64-padBytes {
```
Comment (xiang90) [inline wal/decoder.go:83]: i think we have a max wal size, also a max key-value pair size. i feel these can be set as a better limit.

Comment (gyuho) [reply wal/decoder.go:83]: > i think we have a max wal size, also a max key-value pair size. i feel these can be set as a better limit.

Do we mean define one for max wal entry size? We have request size limit, but I don't see anywhere we define limit? https://github.com/etcd-io/etcd/blob/1166b1f195efae31439c7b3c913b4ef02e7df889/wal/wal.go#L777-L786

Comment (xiang90) [reply wal/decoder.go:83]: we have a request limit. also we know that the throughput wont be unlimited. we probably will only be able to batch at most 1000 requests into one WAL entry at most, usually far less... So I guess a 2GB wal entry limit is large enough...

Comment (gyuho) [reply wal/decoder.go]: Added a check for entry upper limit. Here I assume no project ever allows more than 10 MB raft message entry.

File: wal/wal.go [hunk 2]
```diff
@@ -412,7 +415,12 @@ func (w *WAL) ReadAll() (metadata []byte, state raftpb.HardState, ents []raftpb.
 		case entryType:
 			e := mustUnmarshalEntry(rec.Data)
 			if e.Index > w.start.Index {
-				ents = append(ents[:e.Index-w.start.Index-1], e)
```
Comment (xiang90) [inline wal/wal.go:415]: can we simply check ` 0 <= e.Index-w.start.Index - 1 < len(ents)`?

**Comment by xiang90** [comment]
lgtm

**Pull request merged**